### PR TITLE
Modify mdParser to offer config extension like nuxtjs and expose an hook for customizing of created parser

### PR DESCRIPTION
**Pull request by arrkiin**
To support the [twemoji](https://github.com/twitter/twemoji) output for [markdown-it-emoji](https://github.com/markdown-it/markdown-it-emoji), I had to figure out how to pass the described mapping function to the rules object of markdown-it.

Example from [markdown-it-emoji](https://github.com/markdown-it/markdown-it-emoji):
```
md.renderer.rules.emoji = function(token, idx) {
  return twemoji.parse(token[idx].content);
};
```
So I extended the mdParser so that it receives rules that are specified within the config and passes them to the markdown parser. An example for a config file shows like this:
```
var emoji = require('markdown-it-emoji');
var twemoji = require('twemoji')

module.exports = {
  content: {
    permalink: ':slug',
    page: '/_content',
    generate: [ // for static build
      'get', 'getAll'
    ],
    isPost: false
  },
  parsers: {
    md: {
        use: [
            emoji
        ],
        rules: {
            emoji: function(token, idx) {
                return twemoji.parse(token[idx].content);
            }
        }
    }
  }
}
```

### CODE REVIEW

**Comment by dahrens** [comment]
You are right - currently this lacks flexibility.

But I personally don't like the idea to include parts of markdown-it's API in nuxtents configuration. Next to renderer rules markdown-it also supports inline token rules, block token rules and other things.

Currently `use` is already part of nuxtents configuration and it might be an exception as it makes it possible to load all available plugins. Initially I thought that writing a plugin for markdown-it would be sufficient, but in your case the plugin says: "use the plugin API of markdown-it to change my behaviour"

I'd suggest one of the following approaches.

## Make it optional to pass the whole renderer to nuxtent

From the configuration point of view this might look like this

```javascript
var md = require('markdown-it-emoji');
var emoji = require('markdown-it-emoji');
var twemoji = require('twemoji')

const myMD = md()
myMD.use(emoji)
myMD.renderer.rules['emoji'] = function(token, idx) {
  return twemoji.parse(token[idx].content);
}

// do whatever with the parser

module.exports = {
    // ...
    parsers: {
        md: myMD
    }
}
```

For this solution the plugins that nuxtent currently applies behind the scenes need to be applied on the given instance. This might lead to problems when someone already configured them in front.

## Provide a hook with the constructed

Another option would be to support a hook method that receives the parser constructed by nuxtent.

```javascript
var md = require('markdown-it-emoji');
var emoji = require('markdown-it-emoji');
var twemoji = require('twemoji')

module.exports = {
    // ...
    parsers: {
      md: {
          use: [
              emoji
          ],
          postParserConstruct: function (parser) {
              parser.renderer.rules[emoji] = function(token, idx) {
                  return twemoji.parse(token[idx].content);
              }
          }
      }
    }
}

This one should be quite easy to implement.

**Comment by arrkiin** [comment]
@dahrens, good point. Thank you for your suggestions. Regarding your description, I don't like the downsides of approach number one. Looks too much like a workaround. I think I will make a proposal for your second suggestion. In every software where I want to change or add behaviour, I really like the possibility of using hooks.  This can offer more flexibility in terms of different renderer and configuration options and decouples  the nuxtent configuration from all the other possible configurations.

You are right, the implementation for integrating that hook should be quite simple. I will do a first proposal next week.

**Comment by alidcastano** [comment]
Another option could be to configure it similar to nuxtent modules: 

```
md: {
  use: [
     emoji
   ]
}
```
with rules:  [use: Object, rules: Function]
```
md: {
  plugins: [ 
    [emoji, (token, idx) => {
       return twemoji.parse(token[idx].content);
    }]
  ]
}
```

**Comment by alidcastano** [comment]
Or to get rid of `use` option all together and only expose a hook like you guys mentioned

**Comment by arrkiin** [comment]
@alidcastano, in the first step I introduced the discussed `postConstructionHook`. Indeed it is also possible to integrate all the configuration mechanic of markdown-it in this function hook and therefore eleminate the `use` parameter in `md`. What is your oppinion on this topic @dahrens?

The corresponding configuration for the introduced extension looks like already mentioned by @dahrens.
```
var emoji = require('markdown-it-emoji');
var twemoji = require('twemoji')

module.exports = {
  ...
  parsers: {
    md: {
        use: [
            emoji
        ],
        postConstructionHook: parser => {
            parser.renderer.rules['emoji'] = function(token, idx) {
                return twemoji.parse(token[idx].content);
            }
        }
    }
  }
  ...
}
```

**Comment by dahrens** [comment]
If we throw out `use` it would be straight forward to remove `highlight`, too.

The configuration should than look like that IMO:

```javascript
module.exports = {
  // ...
  parsers: {
    md: parser => {
      // do whatever you want to do with the parser
    }
  }
}
```

I like that as it gives the user all possibilities to do whatever with markdown-it.

@alidcastano what is the problem with appveyor raising errors here? Looks like the fail is not related to this PR, but to the overall configuration.

**Comment by arrkiin** [comment]
@dahrens and @alidcastano, I thought about your ideas of stripping the code in `parser.js`. So I made some consolidations and moved the whole configuration code of markdown-it into an `config` object and integrated the plugin assembly into the new `postConstructionHook`. To check the behaviour I build up a configuration example with all the example-code that I found in the repository. For my local test case everything is working as expected.

```
const Prism = require('prismjs')
const externalLinks = require('markdown-it-link-attributes')
const emoji = require('markdown-it-emoji')
const twemoji = require('twemoji')
const markdownAnchors = require('markdown-it-anchor')

const plugins = [
    [
        markdownAnchors,
        {
            level: 1
        }
    ],
    [
        externalLinks,
        {
            target: '_blank',
            rel: 'noopener'
        }
    ],
    emoji
]

module.exports = {
  ...
  parsers: {
    md: {
        config: {
            preset: 'default',
            html: true,
            typographer: true,
            linkify: true,
            highlight: (code, lang) => {
                return Prism.highlight(
                    code,
                    Prism.languages[lang] || Prism.languages.markup
                )
            }
        },
        postConstructionHook: parser => {
            plugins.forEach(plugin => {
              Array.isArray(plugin)
                ? parser.use.apply(parser, plugin)
                : parser.use(plugin)
            })
            parser.renderer.rules['emoji'] = function(token, idx) {
                return twemoji.parse(token[idx].content);
            }
        }
    }
  }
}
```

**Comment by arrkiin** [comment]
@dahrens, I read your comment a second time and thought about it. Did you imagine a configuration like this?

```
const plugins = [ ... ]
module.exports = {
  ...
  parsers: {
    md: () => {
      const parser = markdownit({...})
      plugins.forEach(plugin => {
        Array.isArray(plugin)
          ? parser.use.apply(parser, plugin)
          : parser.use(plugin)
      })
      parser.renderer.rules['emoji'] = function(token, idx) {
        return twemoji.parse(token[idx].content);
      }
      return parser
    }
  }
}
```

**Comment by dahrens** [comment]
There is currently at least one features in nuxtent that relies on markdown-it to be configured as it is. This is how markdown-it-anchors gets configured based on the anchor level provided in the [nuxtent content configuration](https://nuxtent.now.sh/guide/configuration).

I'd suggest to keep the initial configuration of the parser where it is - but instead of merging this with incoming use parameter I would just provide the `postConstructionHook` to do whatever with the parser.

@alidcastano are there any other parts of the code, that rely on the config passed to markdown-it initially or can we move this configuration as is into the nuxtent configuration?

I'll add commets to the changeset to make this more clear.

File: lib/util/parsers.js [hunk 1]
```diff
@@ -1,31 +1,9 @@
...
 
-  const parser = markdownit({
-    preset: 'default',
-    html: true,
-    typographer: true,
-    linkify: true,
-    highlight
-  })
-
-  const plugins = [
```
Comment (dahrens) [inline lib/util/parsers.js]: This part should stay here.

File: lib/util/parsers.js [hunk 2]
```diff
@@ -1,31 +1,9 @@
...
-    [
-      markdownAnchors,
-      {
-        level: [anchorsLevel]
-      }
-    ]
```
Comment (dahrens) [inline lib/util/parsers.js]: But we must not merge this with incoming use.

File: lib/util/parsers.js [hunk 3]
```diff
@@ -1,31 +1,9 @@
 import yamlit from 'js-yaml'
 import markdownit from 'markdown-it'
-import markdownAnchors from 'markdown-it-anchor'
```
Comment (dahrens) [inline lib/util/parsers.js]: I like the idea to pass the configuration from the outside, but I'm currently not aware of all possible side effects.

**Comment by alidcastano** [comment]
I think it should work similar to how Nuxt uses webpack but allows users to extend it. Nuxtent can use markdown it (provide default configuration along with anchor plugin - yes @dahrens that is the only part that uses it directly).

**Comment by alidcastano** [comment]
Yeah the build is failing due to previous push - will need to look into this soon

**Comment by arrkiin** [comment]
Ok, I reintroduced the default config object, I named it `baseConfig`, for markdown-it and the default plugin `markdownAnchors` with the corresponding `anchorsLevel` parameter. For flexibility purposes the optional `config` parameter overrides the specified configuration options in `baseConfig` with values set in `nuxtent.config.js`.

**Comment by arrkiin** [comment]
During studying the nuxt.js configuration docs I realized, that I could change the mechanism of extending the `config` so that it behaves exactly the same way like the `postConstructionHook`. So this looks more consistent. What is your opinion on this?
```
  parsers: {
    md: {
        configureHook: config => {
            config['highlight'] = (code, lang) => {
                return Prism.highlight(
                    code,
                    Prism.languages[lang] || Prism.languages.markup
                )
            }
        },
        postConstructionHook: parser => {
            plugins.forEach(plugin => {
              Array.isArray(plugin)
                ? parser.use.apply(parser, plugin)
                : parser.use(plugin)
            })
            parser.renderer.rules['emoji'] = (token, idx) => {
                return twemoji.parse(token[idx].content);
            }
        }
    }
```

**Comment by alidcastano** [comment]
Hi @arrkiin thanks for devoting time to this. The last API you proposed seems too complicated.

What do you think of this: 

```
{
  md: {
   extend: Function, 
    plugins: 2D Array - [plugin: Object or Array, options: Object, customRules: Function]
  }
}
```

And can be used as such: 

```
md: { 
 extend (config) {
    config['highlight'] = (code, lang) => {
           return Prism.highlight(
               code,
               Prism.languages[lang] || Prism.languages.markup
            )
        }
  },
  plugins: [
     [[emoji, (parser) => {
          parser.renderer.rules['emoji'] = (token, idx) => {
             return twemoji.parse(token[idx].content);
          }
     }]
 ] 
}
```

The first part is similar to what you had, except it's similar to how Nuxt does it so the API is familiar with users.

The second part, rather than having users initialize plugins themselves, we do it for them, and pass the parser as an extra parameter for advanced customization.

**Comment by arrkiin** [comment]
Sounds good. But looking in your suggestion, I don't exactly figure out how the possible combinations should be distiguished?
1. Plugin without Parameters:
```
plugins = [
  [
    emoji
  ],
  ...
]
```
2. Plugin with Parameters:
```
plugins = [
  [
    [emoji, { ... params ... }]
  ],
  ...
]
```
3. Plugin with extension-hook:
```
plugins = [
  [
    emoji,
    (parser) => {}
  ],
  ...
]
```
4. Plugin with params and extension-hook:
```
plugins = [
  [
    [emoji, { ... params ... }],
    (parser) => {}
  ],
  ...
]
```
Or should we use objects instead of arrays?

**Comment by arrkiin** [comment]
I think I would split the parser-extension-callback off the plugins array. What about this variant? There we have a similar mechanism like the `extend(config)` for customizing the parser (`customize(parser)`) after it was created. It is clean to write and linguistically understandable.
```
  parsers: {
    md: {
        extend(config) {
            config['highlight'] = (code, lang) => {
                return Prism.highlight(
                    code,
                    Prism.languages[lang] || Prism.languages.markup
                )
            }
        },
        plugins: [
            emoji,
            [ externalLinks, { target: '_blank', rel: 'noopener' } ]
        ],
        customize(parser) {
            parser.renderer.rules['emoji'] = (token, idx) => {
                return twemoji.parse(token[idx].content);
            }
        }
    }
  }
```

**Comment by alidcastano** [comment]
API looks good to me :) @dahrens What do you think? 

@arrkiin This PR will have to wait till I can go back and fix build for release

**Comment by dahrens** [comment]
Delicious! ;)

**Pull request merged**